### ISSUE TITLE
It does not move on PHP7.1.5

### ISSUE THREAD

**Issue by git-kurara** [ezyang/htmlpurifier#134]
Hi.
Thank you for a wonderful library.

But it does not move in my environment, I made Issue.

It runs on PHP 7.1.5, but processing will not proceed at the following places.

`library\HTMLPurifier\HTMLModule\List.php:32`
↓
`$ol = $this->addElement('ol', 'List', new HTMLPurifier_ChildDef_List(), 'Common');`

Commenting out this part will work.

Endless run and CPU process is 100%.

Please Help me.

Thank you.

**Comment by git-kurara** [comment]
Add comment .

The problem does not seem to be related to the input that is passed to the purifier. For all input, when I try to purify the text, there are no errors thrown but the process hangs, and the CPU usage for that process becomes 100%.

**Comment by ezyang** [comment]
I ran the test suite with 7.1.5-1+deb.sury.org~xenial+1, but I didn't see any hang. Is this on Windows? Can you test with other versions of PHP and see if there are similar problems?

**Comment by thedotedge** [comment]
Reproducible on 7.1.4 and 7.1.5. Here's the trace from php-fpm:
```
script_filename = /srv/site/www.site.com/bootstrap.php
[0x00007f1c9b213bb0] Composer\Autoload\includeFile() /srv/site/vendor/composer/ClassLoader.php:322
[0x00007f1c9b213b20] loadClass() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/DefinitionCache/Serializer.php:73
[0x00007f1c9b213ac0] spl_autoload_call() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/DefinitionCache/Serializer.php:73
[0x00007f1c9b213a60] unserialize() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/DefinitionCache/Serializer.php:73
[0x00007f1c9b2139e0] get() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/DefinitionCache/Decorator.php:81
[0x00007f1c9b213970] get() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/DefinitionCache/Decorator/Cleanup.php:70
[0x00007f1c9b2138f0] get() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/Config.php:505
[0x00007f1c9b2137d0] getDefinition() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/Config.php:415
[0x00007f1c9b213750] getHTMLDefinition() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier/Generator.php:74
[0x00007f1c9b2136c0] __construct() /srv/site/vendor/ezyang/htmlpurifier/library/HTMLPurifier.php:158
[0x00007f1c9b213540] purify() /srv/site/src/site/Service/MessageService.php:329
[0x00007f1c9b2133b0] send() /srv/site/src/site/Web/Controller/MailboxController.php:170
[0x00007f1c9b2132d0] dopostAction() /srv/site/src/site/Web/Controller/BaseController.php:200
[0x00007f1c9b213240] process() /srv/site/src/site/Web/Dispatcher.php:188
[0x00007f1c9b213120] dispatch() /srv/site/www.site.com/bootstrap.php:61
```

**Comment by git-kurara** [comment]
I environments.

OS : CentOS6.8
PHP : 7.1.5 (Install by yum)

@thedotedge 
Thanks, you are output !!

**Comment by ezyang** [comment]
Thanks @thedotedge. Do you have a small script that causes this? Judging from the trace, it kind of looks like PHP 7 has finally made a backwards breaking change to the serialize() format. If that's the case, running `php maintenance/flush.php` should fix it.

**Comment by thedotedge** [comment]
It was something like:
```
$config = HTMLPurifier_Config::createDefault();
            $config->set('Cache.SerializerPath', '/var/cache/htmlpurifier');
            $config->set('HTML.SafeIframe', true);
            $config->set('HTML.Nofollow', true);
            $config->set('AutoFormat.RemoveSpansWithoutAttributes', true);
            $config->set('Attr.AllowedFrameTargets', ['_blank']); // allow target=_blank
            $config->set('URI.SafeIframeRegexp', '%^(https?:)?//(www\.youtube(?:-nocookie)?\.com/embed/|player\.vimeo\.com/video/)%'); //allow YouTube and Vimeo
            $config->set('HTML.ForbiddenAttributes', ['onload']); // svg onload="alert" XSS
            $htmlPurifier = new HTMLPurifier($config);
           $htmlPurifier->purify($string);
```
I have tried flushing the cache, but then it spiked and 100% exactly at `HTMLPurifier/HTMLModule/List.php:32`

**Comment by ezyang** [comment]
I tried again with php-fpm on Ubuntu but still could not reproduce, so it sounds like this is a CentOS specific problem. Unfortunately, I don't have easy access to any CentOS server; does anyone have one they can lob me ssh access to?

**Comment by fr00x** [comment]
We got hit hard in production today by this as our host updated their PHP Version to 7.1.5.

I can reproduce this loop on a CloudLinux and PHP 7.1.5. Doing the flush Operation has no effect. 

However, on Ubuntu with 7.1.5-1+deb.sury.org~trusty+1 no issues.

I am trying to find a system for you to test.

**Comment by fr00x** [comment]
@git-kurara @thedotedge Is anyone of you guys able to provide a temporary debug/test system for @ezyang?

I can only offer a CentOS 7.3 which in combination with PHP 7.1.5 is not showing any loops right now. The loop only is visible on our production system running CentOS 6.9.

**Comment by MPSinclair** [comment]
Could this be related to opcache?

On CentOS 6.9 with PHP-FPM 7.1.5 and opcache enabled, just including the standalone causes a timeout. I switched to using the lite library and was able to include fine, but trying to run the parser resulted again in timing out. As soon as I disabled opcache the parser executed quickly.

**Comment by ezyang** [comment]
@MPSinclair This wouldn't be the first time an opcache bug caused HTML Purifier to infinite loop, see #108. I'll take a look.

**Comment by ezyang** [comment]
In fact, someone has already filed upstream: https://bugs.php.net/bug.php?id=74623

**Comment by git-kurara** [comment]
@ezyang 

Oh, Thanks.

As a bug in PHP, you have to watch over what can be solved.

**Comment by Xiphin** [comment]
It seems to be fixed by me: https://github.com/Xiphin/htmlpurifier/commit/cb4871f4464659918e0c59b914d359a8874ea62a

**Comment by git-kurara** [comment]
@Xiphin 

Great work !!!!!

It work in my environment.

I hope this Pull Request.

`Fix: using null instead of false. Fixed CPU is 100% on PHP 7.1.*` : https://github.com/ezyang/htmlpurifier/pull/137

**Comment by ezyang** [comment]
I merged. Probably not going to release in the near future.
